packages/contracts/contracts/L1/Colosseum.sol

```diff
@@ -231,86 +313,114 @@ contract Colosseum is Initializable, Semver {
+        // require(ZK_VERIFIER.verify(_proof, _pair), "Colosseum: invalid proof");
 
-        // TODO(chokobole): check transaction root hash
-        // TODO(chokobole): use this public input hash require(_proof[4] == _publicInputHash);
-        bytes32 _publicInputHash = Hashing.hashPublicInput(
-            _srcOutputRootProof.stateRoot,
-            _publicInput,
-            CHAIN_ID,
-            dummyHashes
-        );
+        _validatePublicInput(_srcOutputRootProof, _publicInput, _rlps, _pair[4]);
+        challenge.outputRoot = dstOutputRoot;
+        challenge.current = msg.sender;
+        challenge.next = address(0);
 
-        require(ZK_VERIFIER.verify(_proof, _pair), "Colosseum: invalid proof");
+        // TODO(pangssu): call `createTransaction` to guardian
 
-        L2_ORACLE.deleteL2Outputs(challenge.outputIndex);
-        emit ProofCompleted(challengeId, challenge.outputIndex);
-        _closeChallenge(challengeId, challenge);
+        emit Proven(_outputIndex, dstOutputRoot);
     }
 
-    function asserterTimeout() external {
-        uint256 challengeId = latestChallengeId;
-        Types.Challenge storage challenge = challenges[challengeId];
+    /**
+     * @notice Closes the challenge because the challenger timed out.
+     *         The winner is the validator.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function challengerTimeout(uint256 _outputIndex) external {
+        Types.Challenge storage challenge = challenges[_outputIndex];
         _validateTurn(challenge, ValidateTurnMode.NORMAL);
-        L2_ORACLE.deleteL2Outputs(challenge.outputIndex);
-        _closeChallenge(challengeId, challenge);
+        _closeChallenge(_outputIndex, challenge);
     }
 
-    function challengerTimeout(uint256 _challengeId) external {
-        Types.Challenge storage challenge = challenges[_challengeId];
-        _validateTurn(challenge, ValidateTurnMode.NORMAL);
-        _closeChallenge(_challengeId, challenge);
+    /**
+     * @notice Approve the challenge to replace output.
+     *
+     * @param _outputIndex Index of L2 checkpoint output.
+     */
+    function approveChallenge(uint256 _outputIndex) external {
+        require(msg.sender == GUARDIAN, "Colosseum: sender is not the guardian");
+
+        Types.Challenge storage challenge = challenges[_outputIndex];
+
+        require(
+            _challengeStatus(challenge) == ChallengeStatus.PROVEN,
+            "Colosseum: this challenge is not validated"
+        );
+
+        L2_ORACLE.replaceL2Output(_outputIndex, challenge.segments[0], challenge.current);
+        _closeChallenge(_outputIndex, challenge);
     }
 
+    /**
+     * @notice Reverts if it's not sender's turn.
+     *
+     * @param _challenge The challenge data.
+     * @param _mode      Turn validation mode.
+     */
     function _validateTurn(Types.Challenge storage _challenge, ValidateTurnMode _mode)
```

Comment:
Yes. it makes sense. But `_validateTurn()` also covered checking transaction sender.

Reply:
> proveFault() calls only the challenger

However, it seems that `_validateTurn` do not check if the sender is challenger or not in `proveFault()`.

It seems that `msg.sender` is fixed for all except `bisect()`. Then it seems better to explicitly check `msg.sender`. And for this, instead of updating the challenge's current and next addresses as the turn changes, how about explicitly storing and using the addresses of the challenger and asserter when `creatingChallenge()`?

Reply:
Good. I also think storing addresses by role would help with gas optimization.